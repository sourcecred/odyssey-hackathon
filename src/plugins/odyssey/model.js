// @flow

/**
 * Core "model" logic for the Odyssey plugin.
 * Basically allows creating a data store of priorities, contributions, and people,
 * and compiling that data store into a cred Graph.
 */
import {
  Graph,
  EdgeAddress,
  NodeAddress,
  type NodeAddressT,
} from "../../core/graph";

import * as NullUtil from "../../util/null";

export type OdysseyEntityType = "PERSON" | "PRIORITY" | "CONTRIBUTION";

const NODE_PREFIX = NodeAddress.fromParts(["sourcecred", "odyssey"]);
export const NodePrefix = Object.freeze({
  base: NODE_PREFIX,
  person: NodeAddress.append(NODE_PREFIX, "PERSON"),
  priority: NodeAddress.append(NODE_PREFIX, "PRIORITY"),
  contribution: NodeAddress.append(NODE_PREFIX, "CONTRIBUTION"),
});

export const EdgePrefix = EdgeAddress.fromParts(["sourcecred", "odyssey"]);

export type OdysseyEntity = {|
  +type: OdysseyEntityType,
  +address: NodeAddressT,
  +name: string,
  +description: string,
|};

/**
 * Core data structure backing an Odyssey instance.
 * Right now we identify nodes by human-chosen name.
 * This is very convenient to implement but means that renaming will be a pain.
 * We should probably change it to use a more robust identifier system in the future.
 * (We will also want to give careful thought to how to reconcile multiple simultaneous
 * edits, again in the future.)
 */
export class OdysseyInstance {
  _graph: Graph;
  _addressToEntity: Map<NodeAddressT, OdysseyEntity>;
  _names: Set<string>;

  constructor() {
    this._graph = new Graph();
    this._names = new Set();
    this._addressToEntity = new Map();
  }

  _addNode(
    name: string,
    description: string,
    type: OdysseyEntityType
  ): OdysseyEntity {
    if (this.hasName(name)) {
      throw new Error(`Name conflict: ${name}`);
    }
    this._names.add(name);
    const address = NodeAddress.append(NODE_PREFIX, type, name);
    this._graph.addNode(address);
    const e: OdysseyEntity = {type, name, description, address};
    this._addressToEntity.set(address, e);
    return e;
  }

  // Returns whether the given name is present in the instance.
  // If it is, then no node with this name may be added.
  hasName(name: string): boolean {
    return this._names.has(name);
  }

  *_entitiesOfPrefix(prefix: NodeAddressT): Iterator<OdysseyEntity> {
    for (const a of this._graph.nodes({prefix})) {
      yield NullUtil.get(this._addressToEntity.get(a));
    }
  }

  // Add a new Priority node. May error if this name was ever used before.
  addPriority(name: string, description: string): OdysseyEntity {
    return this._addNode(name, description, "PRIORITY");
  }

  priorities(): Iterator<OdysseyEntity> {
    return this._entitiesOfPrefix(NodePrefix.priority);
  }

  // Add a new Person node. May error if this name was ever used before.
  addPerson(name: string, description: string): OdysseyEntity {
    return this._addNode(name, description, "PERSON");
  }

  people(): Iterator<OdysseyEntity> {
    return this._entitiesOfPrefix(NodePrefix.person);
  }

  // Add a new Contribution node. May error if this name was ever used before.
  addContribution(name: string, description: string): OdysseyEntity {
    return this._addNode(name, description, "CONTRIBUTION");
  }

  contributions(): Iterator<OdysseyEntity> {
    return this._entitiesOfPrefix(NodePrefix.contribution);
  }

  // Cred for these edge flows from src to dst.
  // So if e.g. a contribution was made by a person and supports a priority,
  // there should be an edge from Priority -> Contribution and from
  // Contribution -> Priority.
  addEdge(src: OdysseyEntity, dst: OdysseyEntity): OdysseyInstance {
    const edge = {
      src: src.address,
      dst: dst.address,
      address: EdgeAddress.append(EdgePrefix, src.name, dst.name),
    };
    this._graph.addEdge(edge);
    return this;
  }

  /**
   * Returns the graph underlying this instance.
   * You may use this Graph e.g. to run PageRank on the graph.
   * DO NOT MODIFY THIS GRAPH INDEPENDENTLY.
   * It is owned by the OdysseyInstance.
   * You can always make a copy if you want to make changes.
   * //TODO: Actually make a ReadOnly graph subtype.
   */
  graph(): Graph {
    return this._graph;
  }
}
