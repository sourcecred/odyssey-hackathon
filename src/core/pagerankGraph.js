// @flow

import deepEqual from "lodash.isequal";

import {toCompat, fromCompat, type Compatible} from "../util/compat";
import {
  Graph,
  type Edge,
  type EdgesOptions,
  type NodeAddressT,
  type EdgeAddressT,
  type GraphJSON,
  sortedEdgeAddressesFromJSON,
  sortedNodeAddressesFromJSON,
  NodeAddress,
  type NeighborsOptions,
} from "./graph";
import {
  distributionToNodeDistribution,
  createConnections,
  createOrderedSparseMarkovChain,
  type EdgeWeight,
} from "./attribution/graphToMarkovChain";
import {
  findStationaryDistribution,
  uniformDistribution,
} from "../core/attribution/markovChain";
import * as NullUtil from "../util/null";

export {Direction} from "./graph";
export type {DirectionT, NeighborsOptions} from "./graph";
export type {EdgeWeight} from "./attribution/graphToMarkovChain";
export type EdgeEvaluator = (Edge) => EdgeWeight;

export type ScoredNode = {|
  +node: NodeAddressT,
  +score: number,
|};

export type WeightedEdge = {|
  +edge: Edge,
  +weight: EdgeWeight,
|};

export type ScoredNeighbor = {|
  // The neighbor node, with its score
  +scoredNode: ScoredNode,
  // The edge connecting the target to its neighbor node, with its weight
  +weightedEdge: WeightedEdge,
  // How much score (in absolute terms) was provided to the target by
  // the neighbor node through this weightedEdge
  +scoreContribution: number,
|};

export opaque type PagerankGraphJSON = Compatible<{|
  +graphJSON: GraphJSON,
  // Score for every node, ordered by the sorted node address.
  +scores: $ReadOnlyArray<number>,
  // Weights for every edge, ordered by sorted edge address.
  // We could save the EdgeWeights directly rather than having separate
  // arrays for toWeights and froWeights, but this would lead to an inflated
  // JSON representation because we would be needlessly duplicating the keys
  // "toWeight" and "froWeight" themselves.
  +toWeights: $ReadOnlyArray<number>,
  +froWeights: $ReadOnlyArray<number>,
  +syntheticLoopWeight: number,
|}>;

/**
 * Options to control how PageRank runs and when it stops
 */
export type PagerankConvergenceOptions = {|
  // Maximum number of iterations before we give up on PageRank Convergence
  +maxIterations: number,
  // PageRank will stop running once the diff between the previous iteration
  // and the latest is less than this threshold
  +convergenceThreshold: number,
|};

export type PagerankConvergenceReport = {|
  // A quantitative measure of how close to convergence the final distribution was.
  // Ideally, this value should be near zero.
  // It shows the maximum absolute-valued change of any entry in the distribution
  // if one more Markov action is taken.
  +convergenceDelta: number,
|};

export const DEFAULT_SYNTHETIC_LOOP_WEIGHT = 1e-3;
export const DEFAULT_MAX_ITERATIONS = 255;
export const DEFAULT_CONVERGENCE_THRESHOLD = 1e-7;

const COMPAT_INFO = {type: "sourcecred/pagerankGraph", version: "0.1.0"};

/**
 * PagerankGraph is a wrapper over the Graph class, which adds
 * the ability to run PageRank to compute scores on the Graph.
 *
 * Every node in the underlying Graph is assigned a numerical score in
 * the range [0, 1]. Provided that there are any nodes, the sum of all
 * the scores will be 1 (i.e. the scores are a probability
 * distribution). The scores are assigned by the [PageRank] algorithm;
 * i.e. a node recieves score in proportion to the score of its
 * neighbors. When the PagerankGraph is first constructed, the scores
 * are initialized to a uniform distribution.
 *
 * [PageRank]: https://en.wikipedia.org/wiki/PageRank
 *
 * Every edge in the Graph is assigned an `EdgeWeight`, which includes a
 * `toWeight` (weight from the `src` to the `dst`) and a `froWeight`
 * (weight from the `dst` back to the `src`). Both `toWeight` and
 * `froWeight` must be nonnegative numbers. The weights influence how
 * score flows from node to node. For example, if the node `root` is
 * connected to `a` with a weight of `1` and to `b` with a weight of `2`,
 * then `b` will recieve twice as much score from `root` as `a` does.
 *
 * Every node in the PagerankGraph has an associated `score`. Provided
 * that the graph has at least one node, the scores are a probability
 * distribution over the nodes; i.e. every score is in the range [0,1]
 * and the scores sum to 1.
 *
 * This class is intended to closely mirror the Graph API so as to
 * present a consistent and familiar interface.
 *
 * At present, PagerankGraph does not support any modification to the
 * underlying Graph; doing so will invalidate PagerankGraph and cause
 * its methods to throw errors.
 */
export class PagerankGraph {
  // The Graph backing this PagerankGraph
  _graph: Graph;
  // The score for each Node in the Graph
  _scores: Map<NodeAddressT, number>;
  // The EdgeWeight for each Edge in the Graph
  _edgeWeights: Map<EdgeAddressT, EdgeWeight>;
  // Weight used to connect nodes to themselves, to avoid isolated
  // nodes.
  _syntheticLoopWeight: number;
  // Modification count of the underlying Graph. Used to determine
  // when this PageRankGraph is in an invalid state (due to changes
  // to the graph backing it).
  _graphModificationCount: number;
  // Sum of all outWeights for a node, including the synthetic weight
  _totalOutWeight: Map<NodeAddressT, number>;

  /**
   * Constructs a new PagerankGraph.
   *
   * Note that constructing a PagerankGraph around an empty graph is illegal,
   * as it is impossible to define a probability distribution over zero
   * nodes.
   */
  constructor(
    // The Graph backing this PagerankGraph. Must not be empty.
    graph: Graph,
    // Provides the initial EdgeWeight for every edge
    edgeEvaluator: EdgeEvaluator,
    // The weight we use to connect every node to itself
    // to ensure there are no isolated nodes. Defaults to
    // DEFAULT_SYNTHETIC_LOOP_WEIGHT.
    syntheticLoopWeight: ?number
  ): void {
    if (graph.equals(new Graph())) {
      throw new Error("Cannot construct PagerankGraph with empty graph.");
    }
    this._graph = graph;
    this._graphModificationCount = graph.modificationCount();
    this._syntheticLoopWeight = NullUtil.orElse(
      syntheticLoopWeight,
      DEFAULT_SYNTHETIC_LOOP_WEIGHT
    );
    if (this._syntheticLoopWeight <= 0) {
      throw new Error("syntheticLoopWeight must be > 0");
    }

    // Initialize scores to the uniform distribution over every node
    this._scores = new Map();
    this._totalOutWeight = new Map();
    const graphNodes = Array.from(this._graph.nodes());
    for (const node of graphNodes) {
      this._scores.set(node, 1 / graphNodes.length);
      this._totalOutWeight.set(node, this._syntheticLoopWeight);
    }

    this._edgeWeights = new Map();
    const addOutWeight = (node: NodeAddressT, weight: number) => {
      const previousWeight = NullUtil.get(this._totalOutWeight.get(node));
      const newWeight = previousWeight + weight;
      this._totalOutWeight.set(node, newWeight);
    };
    for (const edge of this._graph.edges()) {
      const weights = edgeEvaluator(edge);
      this._edgeWeights.set(edge.address, weights);
      addOutWeight(edge.src, weights.toWeight);
      addOutWeight(edge.dst, weights.froWeight);
    }
  }

  /**
   * Retrieves the Graph backing this PagerankGraph.
   */
  graph(): Graph {
    this._verifyGraphNotModified();
    return this._graph;
  }

  /**
   * Returns the PagerankGraph's synthetic loop weight.
   *
   * The synthetic loop weight simulates a "phantom loop" connecting
   * every node to itself. This ensures that every node has at least
   * one outgoing connection, so that the corresponding markov chain
   * used for PageRank is well-defined.
   *
   * In general, the synthetic loop weight should be quite small.
   * By default, we set it to 1e-3.
   */
  syntheticLoopWeight(): number {
    return this._syntheticLoopWeight;
  }

  *_nodesIterator(iterator: Iterator<NodeAddressT>): Iterator<ScoredNode> {
    for (const node of iterator) {
      const score = NullUtil.get(this._scores.get(node));
      yield {node, score};
    }
  }

  /**
   * Provides node and score for every node in the underlying graph.
   *
   * Optionally, provide a node prefix to return an iterator containing
   * only node/score objects whose nodes match the provided node prefix.
   * See Graph.nodes and Address.hasPrefix for details.
   */
  nodes(options?: {|+prefix: NodeAddressT|}): Iterator<ScoredNode> {
    this._verifyGraphNotModified();
    const iterator = this._graph.nodes(options);
    return this._nodesIterator(iterator);
  }

  /**
   * Retrieve a node from the graph, along with its score.
   *
   * TODO(#1020): Allow optional filtering, as in Graph.node.
   */
  node(x: NodeAddressT): ?ScoredNode {
    this._verifyGraphNotModified();
    const score = this._scores.get(x);
    if (score == null) {
      return null;
    } else {
      return {node: x, score};
    }
  }

  /**
   * Provides edge and weight for every edge in the underlying graph.
   *
   * Optionally, provide an EdgesOptions parameter to return an
   * iterator containing edges matching the EdgesOptions prefix
   * filter parameters. See Graph.edges for details.
   */
  edges(options?: EdgesOptions): Iterator<WeightedEdge> {
    this._verifyGraphNotModified();
    const iterator = this._graph.edges(options);
    return this._edgesIterator(iterator);
  }

  *_edgesIterator(iterator: Iterator<Edge>): Iterator<WeightedEdge> {
    for (const edge of iterator) {
      const weight = NullUtil.get(this._edgeWeights.get(edge.address));
      yield {edge, weight};
    }
  }

  /**
   * Provides the edge and weight for a particular edge, if present.
   *
   * TODO(#1020): Allow optional filtering, as in Graph.edge.
   */
  edge(a: EdgeAddressT): ?WeightedEdge {
    this._verifyGraphNotModified();
    const edge = this._graph.edge(a);
    if (edge != null) {
      const weight = NullUtil.get(this._edgeWeights.get(edge.address));
      return {edge, weight};
    }
    return null;
  }

  /**
   * Provides the total out weight for a node, i.e. every edge weight pointed
   * away from the node, plus the syntheticLoopWeight.
   *
   * The total out weight is needed to interpret the actual significance of any
   * particular edge's weight, as edge weights are normalized by the totalOutWeight
   * so that the normalized weights going out of a node always sum to 1.
   */
  totalOutWeight(node: NodeAddressT): number {
    this._verifyGraphNotModified();
    const weight = this._totalOutWeight.get(node);
    if (weight == null) {
      throw new Error(
        `Tried to get outWeight for non-existent node ${NodeAddress.toString(
          node
        )}`
      );
    }
    return weight;
  }

  /**
   * Provides the Neighbors to a target node, along with how those
   * neighbors contributed to the node's score.
   *
   * See the docs on `Graph.neighbors` for the semantics of what a `Neighbor`
   * is. This call augments the Neighbors from graph, so that for each neighbor
   * we also have the neighbor node's score, the EdgeWeight for the edge, and a
   * scoreContribution, which shows how much score was contributed to the
   * target node from that Neighbor.
   *
   * When the PagerankGraph is well-converged, it will be the case that a
   * node's score is equal to the score contribution from each neighbor plus
   * the synthetic loop's score contribution.
   *
   * When the PagerankGraph is not well-converged, the score contributions are
   * meaningless.
   */
  neighbors(
    target: NodeAddressT,
    options: NeighborsOptions
  ): Iterator<ScoredNeighbor> {
    this._verifyGraphNotModified();
    if (!this.graph().hasNode(target)) {
      throw new Error(
        `Tried to find neighbors of non-existent node ${NodeAddress.toString(
          target
        )}`
      );
    }
    return this._neighborsIterator(target, options);
  }

  *_neighborsIterator(
    target: NodeAddressT,
    options: NeighborsOptions
  ): Iterator<ScoredNeighbor> {
    const graphNeighbors = this.graph().neighbors(target, options);
    for (const {node, edge} of graphNeighbors) {
      const scoredNode = NullUtil.get(this.node(node));
      const weightedEdge = NullUtil.get(this.edge(edge.address));
      // We compute how much of target's score is attributable to the neighbor.
      // First, we find out how much edge weight there was from node to target,
      // based on whether it was an IN-edge or OUT-edge or loop.
      let relevantEdgeWeight = 0;
      if (edge.src === target) {
        relevantEdgeWeight += weightedEdge.weight.froWeight;
      }
      if (edge.dst === target) {
        relevantEdgeWeight += weightedEdge.weight.toWeight;
      }
      // We normalize this edge weight by the total outWeight for `node`.
      const normalizedEdgeWeight =
        relevantEdgeWeight / this.totalOutWeight(node);

      // Then we directly compute the score contribution
      const scoreContribution = scoredNode.score * normalizedEdgeWeight;
      yield {scoredNode, weightedEdge, scoreContribution};
    }
  }

  /**
   * Returns how much of a node's score came from its synthetic loop.
   * For most nodes, this should be near zero. However, if the node has no
   * outgoing edge edge weight (e.g. it is isolated), then this value
   * may be larger.
   *
   * The results of syntheticLoopScoreContribution are not meaningful if the
   * PagerankGraph is not converged.
   */
  syntheticLoopScoreContribution(node: NodeAddressT): number {
    this._verifyGraphNotModified();
    const scoredNode = this.node(node);
    if (scoredNode == null) {
      throw new Error(
        "Cannot get syntheticLoopScoreContribution for non-existent node"
      );
    }
    return (
      (scoredNode.score * this._syntheticLoopWeight) / this.totalOutWeight(node)
    );
  }

  /**
   * Asynchronously run PageRank to re-compute scores.
   *
   * Calling this method constructs a [Markov Chain] corresponding
   * to the underlying graph and its associated edge weights,
   * and then iteratively converges to the stationary distribution
   * of that chain, according to the [PageRank algorithm].
   *
   * [Markov Chain]: https://brilliant.org/wiki/markov-chains/
   * [PageRank algorithm]: https://en.wikipedia.org/wiki/PageRank
   *
   * The `PagerankConvergenceOptions` gives guidance on how to run
   * PageRank. PageRank will continue running until either
   * `options.maxIterations` has been exceeded, or until the largest
   * individual delta in a node's score between the present and previous
   * iteration is less than or equal to `options.convergenceThreshold`.
   *
   * TODO(#1020): Make `runPagerank` use the current nodes' scores as a
   * starting point for computation, rather than re-generating from
   * scratch every time `runPagerank` is called.
   */
  async runPagerank(
    options: PagerankConvergenceOptions
  ): Promise<PagerankConvergenceReport> {
    this._verifyGraphNotModified();
    const edgeEvaluator = (x: Edge) =>
      NullUtil.get(this._edgeWeights.get(x.address));
    const connections = createConnections(
      this._graph,
      edgeEvaluator,
      this._syntheticLoopWeight
    );
    const osmc = createOrderedSparseMarkovChain(connections);
    const alpha = 0;
    const seed = uniformDistribution(osmc.chain.length);
    const initialDistribution = uniformDistribution(osmc.chain.length);
    const distributionResult = await findStationaryDistribution(
      osmc.chain,
      seed,
      alpha,
      initialDistribution,
      {
        verbose: false,
        convergenceThreshold: options.convergenceThreshold,
        maxIterations: options.maxIterations,
        yieldAfterMs: 30,
      }
    );
    this._scores = distributionToNodeDistribution(
      osmc.nodeOrder,
      distributionResult.pi
    );
    return {
      convergenceDelta: distributionResult.convergenceDelta,
    };
  }

  /**
   * Returns whether another PagerankGraph is equal to this one.
   *
   * PagerankGraphs are considered equal if they have the same nodes with
   * the same scores, and the same edges with the same weights, and the same
   * syntheticLoopWeight.
   *
   * The modification history of the underlying Graph is irrelevant to
   * equality.
   */
  equals(that: PagerankGraph): boolean {
    if (!(that instanceof PagerankGraph)) {
      throw new Error(`Expected PagerankGraph, got ${String(that)}`);
    }
    this._verifyGraphNotModified();
    return (
      this.graph().equals(that.graph()) &&
      deepEqual(this._scores, that._scores) &&
      deepEqual(this._edgeWeights, that._edgeWeights) &&
      this._syntheticLoopWeight === that._syntheticLoopWeight
    );
  }

  /**
   * Serialize this graph into a PagerankJSON object.
   *
   * Returns a plain JavaScript object.
   *
   * For space efficency, we store the node scores as an array of numbers in
   * node-address-sorted order, and we store the edge weights as two arrays of
   * numbers in edge-address-sorted-order.
   */
  toJSON(): PagerankGraphJSON {
    this._verifyGraphNotModified();

    const graphJSON = this.graph().toJSON();
    const nodes = sortedNodeAddressesFromJSON(graphJSON);
    const scores: number[] = nodes.map((x) =>
      NullUtil.get(this._scores.get(x))
    );

    const edgeAddresses = sortedEdgeAddressesFromJSON(graphJSON);
    const edgeWeights: EdgeWeight[] = edgeAddresses.map((x) =>
      NullUtil.get(this._edgeWeights.get(x))
    );
    const toWeights: number[] = edgeWeights.map((x) => x.toWeight);
    const froWeights: number[] = edgeWeights.map((x) => x.froWeight);

    const rawJSON = {
      graphJSON,
      scores,
      toWeights,
      froWeights,
      syntheticLoopWeight: this.syntheticLoopWeight(),
    };

    return toCompat(COMPAT_INFO, rawJSON);
  }

  static fromJSON(json: PagerankGraphJSON): PagerankGraph {
    const {
      toWeights,
      froWeights,
      scores,
      graphJSON,
      syntheticLoopWeight,
    } = fromCompat(COMPAT_INFO, json);
    const graph = Graph.fromJSON(graphJSON);

    const nodes = sortedNodeAddressesFromJSON(graphJSON);
    const scoreMap: Map<NodeAddressT, number> = new Map();
    for (let i = 0; i < nodes.length; i++) {
      scoreMap.set(nodes[i], scores[i]);
    }

    const edges = sortedEdgeAddressesFromJSON(graphJSON);
    const edgeWeights: Map<EdgeAddressT, EdgeWeight> = new Map();
    for (let i = 0; i < edges.length; i++) {
      const toWeight = toWeights[i];
      const froWeight = froWeights[i];
      edgeWeights.set(edges[i], {toWeight, froWeight});
    }

    function evaluator(e: Edge): EdgeWeight {
      return NullUtil.get(edgeWeights.get(e.address));
    }

    const prg = new PagerankGraph(graph, evaluator, syntheticLoopWeight);
    // TODO(#1020): It's a little hacky to force the scores in like this;
    // consider adding an optional constructor argument to allow manually
    // setting the scores at construction time, if we ever find a use case
    // that needs it.
    prg._scores = scoreMap;
    return prg;
  }

  _verifyGraphNotModified() {
    if (this._graph.modificationCount() !== this._graphModificationCount) {
      throw new Error(
        "Error: The PagerankGraph's underlying Graph has been modified."
      );
    }
  }
}
