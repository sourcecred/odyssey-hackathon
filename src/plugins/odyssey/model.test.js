// @flow

import {OdysseyInstance, EdgePrefix} from "./model";
import deepEqual from "lodash.isequal";
import {EdgeAddress} from "../../core/graph";

describe("plugins/odyssey/model", () => {
  const example = () => {
    const instance = new OdysseyInstance();

    // Add people
    const me = instance.addPerson("me", "the author of the code");
    const myPartner = instance.addPerson(
      "my partner",
      "a wonderful human being"
    );

    // Add some priorities
    const hackathonPriority = instance.addPriority(
      "hackathon",
      "it's what we're working on"
    );
    const testingPriority = instance.addPriority(
      "testing",
      "it's very important"
    );

    // Add contributions
    const odysseyModule = instance.addContribution(
      "odyssey/model.js",
      "some nice looking source code"
    );
    const thisFile = instance.addContribution(
      "odyssey/model.test.js",
      "a very fine test file"
    );

    // Helper function to make it easy to check the edges
    const expectedEdges = [];
    function addEdge(src, dst) {
      instance.addEdge(src, dst);
      expectedEdges.push({src, dst});
    }

    // Relate priorities to other entities
    addEdge(hackathonPriority, testingPriority);
    addEdge(testingPriority, thisFile);
    addEdge(hackathonPriority, odysseyModule);

    // Connect files to their authors
    addEdge(thisFile, me);
    addEdge(odysseyModule, me);

    // Connect files to their src code dependencies
    addEdge(thisFile, odysseyModule);

    // Also, I was supported by my partner
    addEdge(me, myPartner);

    return {
      instance,
      me,
      myPartner,
      hackathonPriority,
      testingPriority,
      thisFile,
      odysseyModule,
      expectedEdges,
    };
  };

  it("can retrieve people", () => {
    const {instance, me, myPartner} = example();
    const people = Array.from(instance.people());
    expect(people).toEqual([me, myPartner]);
  });

  it("can retrieve priorities", () => {
    const {instance, testingPriority, hackathonPriority} = example();
    const priorities = Array.from(instance.priorities());
    expect(priorities).toEqual([hackathonPriority, testingPriority]);
  });

  it("can retrieve contributions", () => {
    const {instance, odysseyModule, thisFile} = example();
    const contributions = Array.from(instance.contributions());
    expect(contributions).toEqual([odysseyModule, thisFile]);
  });

  it("errors if a name is duplicated (even across types)", () => {
    const {instance, me} = example();
    const fail = () => instance.addContribution(me.name, "foo bar");
    expect(fail).toThrowError("Name conflict");
  });

  it("accurately reports which names are available", () => {
    const {instance, me} = example();
    expect(instance.hasName(me.name)).toEqual(true);
    expect(instance.hasName("foo bar baz")).toEqual(false);
  });

  it("graph snapshots as expected", () => {
    const {instance} = example();
    expect(instance.graph().toJSON()).toMatchSnapshot();
  });

  it("adding a duplicate node is an error", () => {
    const {instance, me} = example();
    expect(() => instance.addPerson(me.name, me.description)).toThrowError(
      "Name conflict"
    );
  });

  it("adding a duplicate edge is no problem", () => {
    const {instance, me, myPartner} = example();
    const g1 = instance.graph().copy();
    instance.addEdge(me, myPartner);
    expect(instance.graph().equals(g1)).toBe(true);
  });

  it("all edges are as expected", () => {
    const {instance, expectedEdges} = example();
    const allEdges = Array.from(instance.graph().edges());
    expect(allEdges).toHaveLength(expectedEdges.length);
    for (const {src, dst} of expectedEdges) {
      const matchingEdge = {
        address: EdgeAddress.append(EdgePrefix, src.name, dst.name),
        src: src.address,
        dst: dst.address,
      };
      const actualIndex = allEdges.findIndex((x) => deepEqual(x, matchingEdge));
      expect(actualIndex).not.toBe(-1);
    }
  });
});
