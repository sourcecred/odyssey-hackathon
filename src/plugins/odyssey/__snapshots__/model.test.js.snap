// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/odyssey/model graph snapshots as expected 1`] = `
Array [
  Object {
    "type": "sourcecred/graph",
    "version": "0.4.0",
  },
  Object {
    "edges": Array [
      Object {
        "address": Array [
          "sourcecred",
          "odyssey",
          "hackathon",
          "odyssey/model.js",
        ],
        "dstIndex": 0,
        "srcIndex": 4,
      },
      Object {
        "address": Array [
          "sourcecred",
          "odyssey",
          "hackathon",
          "testing",
        ],
        "dstIndex": 5,
        "srcIndex": 4,
      },
      Object {
        "address": Array [
          "sourcecred",
          "odyssey",
          "me",
          "my partner",
        ],
        "dstIndex": 3,
        "srcIndex": 2,
      },
      Object {
        "address": Array [
          "sourcecred",
          "odyssey",
          "odyssey/model.js",
          "me",
        ],
        "dstIndex": 2,
        "srcIndex": 0,
      },
      Object {
        "address": Array [
          "sourcecred",
          "odyssey",
          "odyssey/model.test.js",
          "me",
        ],
        "dstIndex": 2,
        "srcIndex": 1,
      },
      Object {
        "address": Array [
          "sourcecred",
          "odyssey",
          "odyssey/model.test.js",
          "odyssey/model.js",
        ],
        "dstIndex": 0,
        "srcIndex": 1,
      },
      Object {
        "address": Array [
          "sourcecred",
          "odyssey",
          "testing",
          "odyssey/model.test.js",
        ],
        "dstIndex": 1,
        "srcIndex": 5,
      },
    ],
    "nodes": Array [
      Array [
        "sourcecred",
        "odyssey",
        "CONTRIBUTION",
        "odyssey/model.js",
      ],
      Array [
        "sourcecred",
        "odyssey",
        "CONTRIBUTION",
        "odyssey/model.test.js",
      ],
      Array [
        "sourcecred",
        "odyssey",
        "PERSON",
        "me",
      ],
      Array [
        "sourcecred",
        "odyssey",
        "PERSON",
        "my partner",
      ],
      Array [
        "sourcecred",
        "odyssey",
        "PRIORITY",
        "hackathon",
      ],
      Array [
        "sourcecred",
        "odyssey",
        "PRIORITY",
        "testing",
      ],
    ],
  },
]
`;
